### PR TITLE
Add hex data input to send screen

### LINKED ISSUE #3430: Data field is missing in Send screen
Where is data field?

screen shot 2018-03-05 at 06 27 08

### PR DESCRIPTION
**Pull request by whymarrh**
Fixes #3430

This PR adds the ability to edit the hex data for a tx when sending it.

### CODE REVIEW

File: ui/app/components/send/send-content/send-hex-data-row/send-hex-data-row.component.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  };
+
+  static contextTypes = {
+    t: PropTypes.func,
+  };
+
+  shouldComponentUpdate (nextProps, nextState, nextContext) {
+    const {data, inError} = this.props
+    const {nextData, nextInError} = nextProps
+    return data !== nextData || inError !== nextInError
```
Comment (whymarrh) [inline ui/app/components/send/send-content/send-hex-data-row/send-hex-data-row.component.js]: This `shouldComponentUpdate` isn't needed because these are primitive values

File: ui/app/components/send/send-content/send-hex-data-row/send-hex-data-row.component.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    event.target.value = event.target.value.replace(/\n/g, '')
+    updateSendHexData(event.target.value)
+  }
+
+  render () {
+    const {inError} = this.props
+
+    return (
+      <SendRowWrapper
+        label={`${this.context.t('hexData')}:`}
```
Comment (whymarrh) [inline ui/app/components/send/send-content/send-hex-data-row/send-hex-data-row.component.js]: Destructure `this.context`

File: ui/app/components/send/send-footer/send-footer.utils.js [hunk 3]
```diff
@@ -36,9 +36,21 @@ function constructUpdatedTx ({
...
 }) {
+  const unapprovedTx = unapprovedTxs[editingTransactionId]
   const editingTx = {
-    ...unapprovedTxs[editingTransactionId],
-    txParams: addHexPrefixToObjectValues({ from, gas, gasPrice }),
+    ...unapprovedTx,
+    txParams: Object.assign(
+      unapprovedTx.txParams,
+      addHexPrefixToObjectValues({data}),
+      addHexPrefixToObjectValues({data: unapprovedTx.txParams.data}),
```
Comment (whymarrh) [inline ui/app/components/send/send-footer/send-footer.utils.js]: This isn't particularly clear and needs explanation—why is this needed?

File: ui/app/components/send/send-content/send-hex-data-row/send-hex-data-row.component.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  shouldComponentUpdate (nextProps, nextState, nextContext) {
+    const {data, inError} = this.props
+    const {nextData, nextInError} = nextProps
+    return data !== nextData || inError !== nextInError
+  }
+
+  onInput = (event) => {
+    const {updateSendHexData} = this.props
+    event.target.value = event.target.value.replace(/\n/g, '')
+    updateSendHexData(event.target.value)
```
Comment (whymarrh) [inline ui/app/components/send/send-content/send-hex-data-row/send-hex-data-row.component.js]: If value is an empty string we should be removing the data

**Review by alextsg — APPROVED** [review]
